examples/platforms/samr21/radio.c

```diff
@@ -0,0 +1,611 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread platform abstraction for radio communication.
+ *
+ */
+
+#include "asf.h"
+
+#include <openthread/config.h>
+#include <openthread/platform/alarm-milli.h>
+#include <openthread/platform/radio.h>
+
+#include "platform-samr21.h"
+#include "common/logging.hpp"
+
+enum
+{
+    IEEE802154_ACK_LENGTH      = 5,
+    IEEE802154_FCS_SIZE        = 2
+};
+
+static otRadioFrame sTransmitFrame;
+static uint8_t      sTransmitPsdu[OT_RADIO_FRAME_MAX_SIZE];
+
+static otRadioFrame sReceiveFrame;
+
+static bool         sSleep          = false;
+static bool         sRxEnable       = false;
+static bool         sTxDone         = false;
+static bool         sRxDone         = false;
+static otError      sTxStatus       = OT_ERROR_NONE;
+static int8_t       sPower          = OPENTHREAD_CONFIG_DEFAULT_MAX_TRANSMIT_POWER;
+static otRadioState sState          = OT_RADIO_STATE_DISABLED;
+static bool         sPromiscuous    = false;
+static uint8_t      sChannel        = 0xFF;
+
+static int8_t       sMaxRssi;
+static uint32_t     sScanStartTime;
+static uint16_t     sScanDuration;
+static bool         sStartScan      = false;
+
+static int8_t         sTxPowerTable[] = { 4, 4, 3, 3, 3, 2, 1, 0,
+                                          -1, -2, -3, -4, -6, -8, -12, -17
+                                        };
+
+/*******************************************************************************
+ * Static
+ ******************************************************************************/
+
+static void radioSleep()
+{
+    if (!sSleep)
```

Comment:
Should this be one of the `OT_EXPECT` functions rather then an `if` surrounding the entire function?

Reply:
I would use return (`if (sleep) return;`) in this case but it will brake coding style rule. 
IMHO such a project specific macro makes code hard to read. So, either "surrounding if" (which looks not so good either) or OT_EXPECT. Respecting also existing rule, OT_EXPECT would be more suitable I guess.

Reply:
Yea, we tend to stick with the `OT_EXPECT` macros since they limit the number of full function `if` statements, making the code more concise.  It's essentially the same as `if (sleep) return;` since it is actually just a `goto` statement.

Reply:
I have a doubt about using otEXPECT in this case and below. It looks like otEXPECT was designed to check states, error conditions etc. (according to its comment: `This checks for the specified condition, which is expected to commonly be true, and branches to the local label "exit" if the condition is false.`). In this case `sSleep` just prevent going into sleep few times in a row. But I'm Ok to put otEXPECT if the maintainer approves.

Reply:
@abtink what do you think?

Reply:
Personally, I think both ways are perfectly fine (I'd consider this a sw developer's choice).